torch/nn/modules/module.py

```diff
@@ -575,11 +576,29 @@ def __getattr__(self, name):
         if '_buffers' in self.__dict__:
             _buffers = self.__dict__['_buffers']
             if name in _buffers:
+                # Needs to recompute the parametrization if the computation graph needs to be computed
+                # This may happen if the parametrization was computed within a torch.no_grad() clause.
+                parametrizations = parametrize.get_auto_parametrizations(self, name)
+                if (parametrizations
+                    and not _buffers[name].grad_fn
+                    and torch.is_grad_enabled()
+                ):
+                    parametrize.update_auto_parametrization(self, name, parametrizations)
                 return _buffers[name]
+
         if '_modules' in self.__dict__:
             modules = self.__dict__['_modules']
             if name in modules:
                 return modules[name]
+
+        parametrizations = parametrize.get_auto_parametrizations(self, name)
+        # If it is a parametrized tensor, we generate it
+        if parametrizations:
+            # We generate the self._buffers[name] attribute
+            parametrize.update_auto_parametrization(self, name, parametrizations)
+            # ... and return it
+            return self.__dict__['_buffers'][name]
```

Comment:
I discussed this with @albanD and he was happy with this. Another way we discussed would be to monkey-patch the method on the fly, but this has problems of its own. It'd probably be best to discuss this with the maintainers of `module.py`, as you suggest.

Reply:
cc @soumith @apaszke @gchanan could we get your opinions on this change to the `nn.Module` base class?
In particular, this only changes the `__getattr__()` function to properly handle re-parametrized parameters by recomputing them if needed.

Reply:
The justification to modify the core module are:
- we want this to be supported for all the modules provided in the core library.
- as it is a core feature, we shouldn't rely on monkey-patching the nn.Module functions which will lead to even more side effects and composability issues.

The overhead for `__getattr__` should be fairly small: a function call and 2 gets on a dictionary. You think this is too much?

Reply:
Ok, so the reason why I get very nervous when I see things like that in the core is that special-casing our own code generally means that we think that we're more important than our users for some reason. When we're not 100% convinced that what we're building will cover all their possible needs (and I'm not sure in this case), then instead of providing one built-in solution which covers some subset of their use-cases we should focus on enabling extension points that really let them solve the problem in full generality by providing the code they want. Note that `nn.Parameter` is a good exception to this rule because it's simply a nice syntax sugar for `nn.Modules`, but it can be implemented in a different way too (i.e. it doesn't restrict anything). This is different in the sense that by including this `if` you're special casing the behavior which cannot be implemented otherwise or customized. One particular limitation present here, which might make this mechanism not work in all cases, is how do you decide when to recompute those parameters. There's no single good answer to that meaning that we should never hard-code an arbitrary choice in a core module.

Regarding the performance, you'd be surprised how much all of those things can cost. `__getattr__` is absolutely critical for performance of `nn.Module`s as it's called every time you access a submodule or a parameter (which happens multiple times in every line). Even very small slowdowns are noticeable in end-to-end training times.